sympy/functions/combinatorial/numbers.py

```diff
@@ -2507,15 +2507,15 @@ def ok(x):
             return S.Zero
         if type(i) is not dict:
             s = list(i)
-            ms = multiset(s)
+            ms = dict(Counter(s).items())
         elif type(i) is dict:
```

Comment:
If `multiset` is going to be deprecated then routines that were looking for `dict` need to be made to look for `Counter`, e.g. `multiset_combinations`, `multiset_permutations`, etc...

Reply:
Ok, perfect. I'll do it in the following days. Is everyone fine with this change?

Reply:
I don't have a strong preference, but I do prefer retaining the original name and would be more in favor of `multiset` returning a `Counter`. What is the preference of others?

Reply:
I personally don't really like `Counter` as a multiset object. It supports some set operations, but not all. For instance, it supports `|`, `&`, and `-` but not the methods `intersection`. And it also doesn't support mixing `Counter` and `set` in set operations. To me a `Counter` is like a multiset, but isn't really one as it doesn't actually support operations like one.

My preference would be to remove (or deprecate) the multiset function and replace its uses with a direct `dict(Counter(...))` call. And if we really want it we can add a native FiniteMultiSet object that works like a real mathematical multiset.

Reply:
Although I will say that code like this line should be using `isinstance` regardless. `type(i) is ...` is bad practice in general.

Reply:
@asmeurer , are you imagining a new proposal for `multiset` object? Because the one we have is just a simple dict right now (which doesn't support `intersection`.

Why don't we just keep this first step simple: let `multiset` return `dict(Counter(dat))` so you get whatever advantages there are for using a native Python object to do the work that is currently done by the function.

Reply:
Im saying that *if* we want to upgrade the API to return something more sophisticated then dict it should be a real multiset object. Such a thing would be in scope for the sets module, but for now I think we should just continue to return a dict.

Reply:
Ok, i changed `multiset` to use a Counter internally and reverted all  the other changes.